src/strkey.rs

```diff
@@ -82,17 +98,134 @@ impl StrkeyPrivateKeyEd25519 {
+
+impl StrkeyMuxedAccountEd25519 {
+    pub fn to_string(&self) -> String {
+        encode(version::MUXED_ACCOUNT_ED25519, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        match payload.try_into() {
+            Ok(muxed) => Ok(Self(muxed)),
+            Err(_) => Err(DecodeError::Invalid),
+        }
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::MUXED_ACCOUNT_ED25519 => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+pub struct StrkeyHashTx(pub [u8; 32]);
+
+impl StrkeyHashTx {
+    pub fn to_string(&self) -> String {
+        encode(version::HASH_TX, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        match payload.try_into() {
+            Ok(hash_tx) => Ok(Self(hash_tx)),
+            Err(_) => Err(DecodeError::Invalid),
+        }
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::HASH_TX => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+pub struct StrkeyHashX(pub [u8; 32]);
+
+impl StrkeyHashX {
+    pub fn to_string(&self) -> String {
+        encode(version::HASH_X, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        match payload.try_into() {
+            Ok(hash_x) => Ok(Self(hash_x)),
+            Err(_) => Err(DecodeError::Invalid),
+        }
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::HASH_X => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Hash, PartialEq, Eq, Debug)]
+// The largest strkey is a signed payload:
+// 32-byte public key + 4-byte payload length + 64-byte payload
+pub struct StrkeySignedPayloadEd25519(pub Vec<u8>);
+
+impl StrkeySignedPayloadEd25519 {
+    pub fn to_string(&self) -> String {
+        encode(version::SIGNED_PAYLOAD_ED25519, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        match payload.try_into() {
+            Ok(signed_payload) => {
+                let payload_len = payload.len() as u32;
+                if payload_len < 32 + 4 + 4 || payload_len > 32 + 4 + 64 {
+                    return Err(DecodeError::Invalid);
+                }
+                let inner_payload_len =
+                    u32::from_be_bytes((&payload[32..32 + 4]).try_into().unwrap());
+                if inner_payload_len + ((4 - inner_payload_len % 4) % 4) != payload_len - 32 - 4 {
+                    return Err(DecodeError::Invalid);
+                }
```

Comment:
I suspect we could write this using an iterator or a cursor so you can read chunks reducing the number of constant values that need to be in this code.

Reply:
I tried, but I think the code written using iterator or cursor is more complicated than that, maybe I didn't find the right path. 😔